Add SiPM spectra city

andLaing opened the pull request:
Addition of city and related functions for the generation of SiPM
photoelectron spectra (binned pdfs).

Additions include:

1) invisible_cities/cities/sipmPDF.py (and corresponding test) which
   reads raw data and generates two types of spectra for each channel
   once converted into pe (zeroed and in pe and zeroed, in pe and
   difference from mau);

2) invisible_cities/io/hist_io.py, a histogram writer to save on
   output data size.

Changes to existing functions include:

+ change to invisible_cities/peak_functions_c.pyx.signal_sipm to
  include calibration option so that all bin data is saved rather than
  just selected data

  + cal=0 (default) is equivalent to unchanged function,
  + cal=1 saves sipmrawf-mean,
  + cal=2 saves sipmrwf-mean-mau

Undid accidental include of some changes from an unrelated
branch.

andLaing committed:
Add SiPM spectra city

Addition of city and related functions for the generation of SiPM
photoelectron spectra (binned pdfs).

Additions include:

1) invisible_cities/cities/sipmPDF.py (and corresponding test) which
   reads raw data and generates two types of spectra for each channel
   once converted into pe (zeroed and in pe and zeroed, in pe and
   difference from mau);

2) invisible_cities/io/hist_io.py, a histogram writer to save on
   output data size.

Changes to existing functions include:

+ change to invisible_cities/peak_functions_c.pyx.signal_sipm to
  include calibration option so that all bin data is saved rather than
  just selected data

  + cal=0 (default) is equivalent to unchanged function,
  + cal=1 saves sipmrawf-mean,
  + cal=2 saves sipmrwf-mean-mau

Undid accidental include of some changes from an unrelated
branch.

jjgomezcadenas reviewed (commented):
I have started to review this PR, probably best to talk in person (on Wednesday) to make faster progress. Meanwhile, please, find a name you like for the city 
https://en.wikipedia.org/wiki/Invisible_Cities
and change the name of your script (naming cities following Calvino's is mandatory in IC :)!

andLaing committed:
Create calib folder.

Created separate folder for 'cities' related
to calibration of sensors.

sipmPDF.py and related test moved to this folder.

jjgomezcadenas committed:
Change name of class from Sipmpdf to SipmPdf.

This is done to conform to convention (hopefully agreed by the team)
of starting class name with Capitals and then use intermediate
capitals for clarity.

andLaing committed:
Changes pythonize and make work again

Fixed error in SipmPdf class name so recognised
when capitalize used.

generalized bin/city so it can search in calib folder

changed names to be more pythonish

added sipm_pdf to command_line_execution_test.py so checked

jacg commented on bin/city:
```diff
@@ -5,15 +5,20 @@ from importlib import import_module
-    exit(1)
-else:
-    city_class.drive(args)
+for i, fldr in enumerate(city_dirs):
```
We use the name 'directory' to refer to directories in IC, so please avoid introducing a different term for it: 'folder'. Please spell it `for dir_ in city_dirs`. The trailing underscore ensures that we avoid shadowing a built-in. Alternatively, `directory` would be fine too.

But, more importantly, is there any good reason why this city can't live with all the other ones? Do we really need to introduce an extra place where cities might live?

This touches on something else that has been bothering me with this city: it seems to introduce modifications to parts of IC where I wouldn't normally expect authors of new cities to have to go. (This is just the impression I get after a cursory look: I haven't had the time to look at it in any detail.)

*If* (and it's a big *if*) the need to add this extra directory turns out to be genuine, then we can do without the `enumerate` and the `if i == len(city_dirs)-1`. The correct spelling of that idea is to cut the old body of the `NoModuleFoundError` handler, replacing it with `pass` or `continue`, and to move it to the `else` block of the `for loop` (yes, the `for loop`!) The addition of `break` to the `else` block of the `try` will still be necessary.

jacg commented on invisible_cities/calib/sipm_pdf.py:
```diff
@@ -22,7 +22,7 @@
 
 
-class SipmPdf(CalibratedCity):
+class Sipm_pdf(CalibratedCity):
```
We conform to PEP8 when it comes to class names (unless there is some overriding reason not to), so this change would have to be reverted, were it not for the fact that it's not a Calvino name, so @jjgomezcadenas is not going to accept the old name either.

andLaing committed:
Reverted names of program/class to approved format

Changed sipm_pdf.py -> sipmpdf.py with equivalent changes
for tests and example config files.

Updated command_line_execution_test.py to reflect the above change.

jjgomezcadenas committed:
Add functions for pedestal subtraction and normalization.

Three methods can be used to compute pedestal and return a normalized
waveform for the SiPMs.

1. Using the mean of the waveform as pedestal (fast)
2. Using a Mau to compute the pedestal (slow)
3. Returning the values above a thredhold (on top of the MAU, also
slow).

jjgomezcadenas committed:
Split peak_functions_c in two files.

peak_functions_c contained functions related with peak finding (as the
name suggests) AND functions related with sensor calibration (e.g,
removing pedestals, divinding by calibration constants). This commit
splits the original file, adding calib_sensor_functions_c, which
collects the sensor-calibration related functions. Tests files are
also split and the changes are propagated to base_cities and
peak_functions.

jjgomezcadenas committed:
Make thr a dynamic variable.

Thr (thresholds) can be used as either single threshold value or as a
vector of individual thresholds (as in the original function by
AL). There is no performance penalty here (among other things because
the function is already very slow)

jjgomezcadenas committed:
Fix bug in sipm_signal_above_trh_mau

Forgotten loop added

jjgomezcadenas committed:
Add tests.

Function sipm_signal_above_thr passes the same tests than original
function (signal_sipm)

jjgomezcadenas committed:
Add new tests.

Two new tests demonstrate that subtracting the baseline from a
waveform distributed around the baseline yields a baseline close to zero.

jjgomezcadenas committed:
Replace signal_sipm by sipm_signal_above_threshold.

The function proposed by AL uses an option (Cal) to perform three
different calculations: If Cal = 0, the function return the values of
the waveform above the baseline + MAU. If Cal = 1, the function
returns the calibrated function (wd - ped) /adc_to_pes, while if Cal=2
it returns the same but adding a MAU. The proposed alternative splits
signal_sipm in three functions, corresponding to the different values
of Cal for clarity. Replacement of signal_sipm by
sipm_signal_above_threshold in peak functions has no effect (the
results of both functions as shown by tests are identical)

jjgomezcadenas committed:
Make signal_sipm private.

The signature of signal_sipm is changed to _signal_sipm, to mark its
new status as private. The function is still used by the tests and by base_cities.

jjgomezcadenas committed:
Tests showing new functions yield same results than old function.

New tests have been added to guarantee that the new functions proposed
for the API yield the same results than the old function (now made private).

jjgomezcadenas reviewed (approved):
As promised, since no body else has complained, I consider that Andrew's first city can be approved. This PR has triggered a lot of discussion, both on performance and on city structure! Bravo!

@gonzaponte, @abotas or @jmbenlloch, please merge
